src/_pytest/fixtures.py

```diff
@@ -464,12 +485,17 @@ def _getnextfixturedef(self, argname: str) -> "FixtureDef[Any]":
             assert self._pyfuncitem.parent is not None
             parentid = self._pyfuncitem.parent.nodeid
             fixturedefs = self._fixturemanager.getfixturedefs(argname, parentid)
-            # TODO: Fix this type ignore. Either add assert or adjust types.
-            #       Can this be None here?
-            self._arg2fixturedefs[argname] = fixturedefs  # type: ignore[assignment]
-        # fixturedefs list is immutable so we maintain a decreasing index.
+            if fixturedefs is not None:
+                self._arg2fixturedefs[argname] = fixturedefs
+        # No fixtures defined with this name.
+        if fixturedefs is None:
+            raise FixtureLookupError(argname, self)
+        # The are no fixtures with this name applicable for the function.
+        if not fixturedefs:
+            raise FixtureLookupError(argname, self)
         index = self._arg2index.get(argname, 0) - 1
-        if fixturedefs is None or (-index > len(fixturedefs)):
+        # The fixture requested its own name, but no remaining to override.
+        if -index > len(fixturedefs):
             raise FixtureLookupError(argname, self)
```

Comment:
we should either merge the conditions or give a extra indication of the difference between not found, not applicable and no override avaliable

i'd prefer to add extra inication

Reply:
I agree, I split the conditions to make the distinct cases clearer. But `FixtureLookupError` does some complex message formatting, it will require a bit of effort to pass the info along cleanly. I might do it in the future.

Reply:
how about putting in some helpfull text constants right now as 3rd parameter and then following up with more intentional formatting whenever

Reply:
I'm happy to attempt this in a follow up, in case @bluetech prefers to move this along.

Reply:
I'm happy to leave this for a follow-up as well

Reply:
@nicoddemus I'll be happy if you try it, it will require some changes in `FixtureLookupError` . (Note that `FixtureLookupError` if exported and its ctor is not doing the `_ispytest` check, though I doubt there are plugins with instantiate it directly).

`FixtureLookupError` already tries to handle the last case directly "recursive dependency involving fixture 'fix' detected", though it will be nice to distinguish between the first ("not found") and second ("not visible to this test") cases.

Reply:
Gave this a quick try, but the result was not much better and it became a bit more complex I'm afraid.